accepted/2020/minimum-os-version/minimum-os-version.md

```diff
@@ -0,0 +1,246 @@
+    string platformIdentifier,
+    int majorVersion, int minorVersion, int servicingVersion, int bindingRevision
+    string message
+    ) {}
+    public AvailabilityKind AvailabilityKind { get; }
+    public string Message { get; }
+    public string PlatformIdentifier { get; }
+    public Version OSVersion { get; }
+    public int BindingRevision { get; }
+}
+
+public enum AvailabilityKind
+{
+    Introduced,
+    Obsoleted,
+    Removed
+}
+```
+
+The platform string is expected to have the same values as used in the `TargetPlatformIdentifier` portion of platform-specific TFMs. Having it in the availability attributes may seem redundant as it will also be burned into assembly attributes for platform-specific reference assemblies. However, explicitly specifying the attribute is useful for platform-agnostic `net5.0` reference assemblies that have methods that are only available on some platforms e.g. due to having platform-specific implementations via bait-and-switch. This may also improve source readability when multi-targeting as it will remove the need to use `#if` around the annotations.
+
+Constants for the platform identifiers could perhaps be exposed on a `PlatformIdentifier` type.
+
+The values declared on this attribute are expected to be accessible by extracting constructor arguments from metadata. However, for readability it’s expected to be applied using subclasses with more specific names. To enforce this, it has an internal constructor and its subclasses are sealed.
+
+The concrete subclasses are `IntroducedAttribute`, `ObsoletedAttribute` and `RemovedAttribute` subclasses, which take of the following form:
+
+```csharp
+public sealed class IntroducedAttribute : AvailabilityAttribute
+{
+    public IntroducedAttribute(
+    string platform,
+    int majorVersion,
+    int minorVersion = 0,
+    int servicingVersion = 0,
+    int bindingRevision = 0,
+    string message = null
+    ) : base(platform, majorVersion, minorVersion, servicingVersion, bindingRevision, message) {}
+}
+```
+
+These attributes are used as follows:
+
+```csharp
+[Introduced(PlatformIdentifier.macOS, 10, 2)]
+[Deprecated(PlatformIdentifier.macOS, 10, 9)]
+[Removed(PlatformIdentifier.macOS, 10, 12, 1)]
+public class Foo : NSObject { ... }
+```
+
+The purpose of keeping ‘removed’ APIs and annotating them with `RemovedAttribute` is to allow apps and libraries to continue to use these APIs in fallback paths for older OS versions. This is a pattern that exists on Apple platforms.
+
+If and when C# support target typing for enums, we could add convenience overloads that would improve readability:
+
+```csharp
+[Introduced (.macOS, 10, 1)]
+[Obsoleted (.macOS, 10, 8)]
+[Removed (.macOS, 10, 12)]
+```
+
+> NOTE: These attributes would only be needed on reference assemblies. They would not be needed at runtime. They could be removed by the linker, and we could also strip them from the implementation assemblies that we ship ourselves.
```

Comment:
I don't think we even need to do that. The attributes will exist in reference assemblies only which are not processed by illinker.

Reply:
For the BCL itself it has not impact, but for platform assemblies (our main use case) we need the attributes in the implementation assemblies since they are heavily used by our test suite.

Reply:
I think that's a very specific use case and size increase does not matter in this case. I also don't think everyone should pay the penalty of having the attributes in implementation libs just because it makes your testing simpler.

Reply:
Added a note, but I'm not sure how we could make it easy to omit these from reference assemblies, though.